src/VirtoCommerce.CatalogModule.Core/Model/DynamicAssociations/DynamicAssociationEvaluationContextExtensions.cs

```diff
@@ -0,0 +1,89 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using VirtoCommerce.Platform.Core.Common;
+
+namespace VirtoCommerce.CatalogModule.Core.Model.DynamicAssociations
+{
+    public static class DynamicAssociationEvaluationContextExtensions
+    {
+        #region DynamicAssociationEvaluationContext extensions
+
+        public static bool IsItemInCategory(this DynamicAssociationEvaluationContext context, string[] categoryIds, string[] excludingCategoryIds, string[] excludingProductIds)
+        {
+            var result = context.Products.InCategories(categoryIds)
+                               .ExcludeCategories(excludingCategoryIds)
+                               .ExcludeProducts(excludingProductIds)
+                               .Any();
+            return result;
+        }
+
+        public static bool AreItemPropertyValuesEqual(this DynamicAssociationEvaluationContext context, Dictionary<string, string> propertyValues)
+        {
+            var result = context.Products.WithPropertyValues(propertyValues).Any();
+
+            return result;
+        }
+
+        #endregion DynamicAssociationEvaluationContext extensions
+
+        #region CatalogProduct extensions
+
+        public static IEnumerable<CatalogProduct> ExcludeCategories(this IEnumerable<CatalogProduct> products, string[] categoryIds)
+        {
+            var retVal = products.Where(x => !ProductInCategories(x, categoryIds));
+            return retVal;
+        }
+
+        public static IEnumerable<CatalogProduct> ExcludeProducts(this IEnumerable<CatalogProduct> products, string[] productIds)
+        {
+            var retval = products.Where(x => !ProductInProducts(x, productIds));
+            return retval;
+        }
+
+        public static IEnumerable<CatalogProduct> InCategories(this IEnumerable<CatalogProduct> products, string[] categoryIds)
+        {
+            categoryIds = categoryIds.Where(x => x != null).ToArray();
+            var productArray = products as CatalogProduct[] ?? products.ToArray();
```

Comment:
Why do you made cast here?

Reply:
Just the existing code copy, will change.